### PR TITLE
Add keyboard controls to the dropdown

### PR DESCRIPTION
**Pull request by hututudor**
https://etvas.atlassian.net/browse/ETV-5810

### CODE REVIEW

File: src/Dropdown/Dropdown.jsx [hunk 1]
```diff
@@ -171,6 +177,76 @@ const Dropdown = ({
...
+
+      if (e.code === 'ArrowUp') {
+        e.preventDefault()
+
+        setCurrentKeyboardFocus(hover =>
+          Math.min(options.length - 1, Math.max(-1, hover - 1))
+        )
+      }
+
+      if (e.code === 'Enter') {
```
Comment (sergiuelisei) [inline src/Dropdown/Dropdown.jsx:208]: A switch statement would be a better fit for this

File: src/Dropdown/Dropdown.jsx [hunk 2]
```diff
@@ -171,6 +177,76 @@ const Dropdown = ({
...
+
+    if (multiple) {
+      return
+    }
+
+    const optionIndex = options.findIndex(option =>
+      isEqual(option.props.value, value)
+    )
+
+    setCurrentKeyboardFocus(optionIndex)
```
Comment (sergiuelisei) [inline src/Dropdown/Dropdown.jsx:247]: I think we can skip this `.find` and use `setCurrentKeyboardFocus(optionIndex)` inside of the children(`<Option {...existingProps} index={index} />`) exactly when we are selecting the option

File: src/Dropdown/Option.jsx [hunk 3]
```diff
@@ -47,13 +53,21 @@ const DropdownItem = ({
     </Flex>
   )
 
+  useLayoutEffect(() => {
+    if (hasKeyboardFocus) {
+      optionRef.current.scrollIntoView({ block: 'nearest', inline: 'nearest' })
```
Comment (sergiuelisei) [inline src/Dropdown/Option.jsx:58]: It looks like we have partial support for `scrollIntoView` on safari, please test this
https://caniuse.com/?search=scrollIntoView

An alternative would be to calculate and set the list container scroll position. I think we can have everything we need (list client height, list scroll height, option height)

File: src/Dropdown/Dropdown.jsx [hunk 4]
```diff
@@ -171,6 +177,76 @@ const Dropdown = ({
...
+        }
+      }
+    }
+
+    const removeCurrentHover = () => {
+      setCurrentKeyboardFocus(-1)
+    }
+
+    const element = wrapper.current
+    element.addEventListener('keydown', handleKeyDownEvent)
```
Comment (sergiuelisei) [inline src/Dropdown/Dropdown.jsx:225]: If you have a reference, you can use the `onKeyDown` prop on the target
`onKeyDown={handleKeyDownEvent}`

**Review by sergiuelisei — CHANGES_REQUESTED** [review]

File: src/Dropdown/Dropdown.jsx (hunk 4 above)
Comment (hututudor) [reply src/Dropdown/Dropdown.jsx]: 8c6f23348e959e8dd428f3991cfb3f5652b572e7

File: src/Dropdown/Dropdown.jsx (hunk 1 above)
Comment (hututudor) [reply src/Dropdown/Dropdown.jsx]: 8c6f23348e959e8dd428f3991cfb3f5652b572e7

File: src/Dropdown/Dropdown.jsx (hunk 2 above)
Comment (hututudor) [reply src/Dropdown/Dropdown.jsx:241]: This is used when we open an already selected dropdown and we need to set the current option index to the selected one. It does not make sense to move that responsibility in the children, especially when we only need this to run when the dropdown opens

File: src/Dropdown/Dropdown.jsx [hunk 5]
```diff
@@ -171,6 +177,70 @@ const Dropdown = ({
     [disabled, error, isCollapsed]
   )
 
+  const removeCurrentKeyboardFocus = useCallback(() => {
+    setCurrentKeyboardFocus(-1)
+  }, [])
+
+  const handleDropdownKeyDown = useCallback(
```
Comment (alsofronie) [inline src/Dropdown/Dropdown.jsx]: Why not only `handleKeyDown`? We're in the Dropdown context/file.

File: src/Dropdown/Option.jsx (hunk 3 above)
Comment (hututudor) [reply src/Dropdown/Option.jsx:58]: The partial support refers to `smooth` scroll behaviour. We don't use it so it's not going to be a problem.

File: src/Dropdown/Dropdown.jsx (hunk 5 above)
Comment (hututudor) [reply src/Dropdown/Dropdown.jsx]: eb1464625c16f68cd887b40b4075fef822e0679e

File: src/Dropdown/Option.jsx (hunk 3 above)
Comment (alsofronie) [reply src/Dropdown/Option.jsx:58]: The partial support only refers to smooth scrolling option, we're in the green here.

File: src/Dropdown/Dropdown.jsx [hunk 6]
```diff
@@ -251,6 +324,31 @@ const Dropdown = ({
...
+
+  if (keys1.length !== keys2.length) {
+    return false
+  }
+
+  for (let key of keys1) {
+    if (object1[key] !== object2[key]) {
+      return false
+    }
+  }
```
[the remark this answers is not in the document]
Comment (hututudor) [reply src/Dropdown/Dropdown.jsx]: 32423c243ed61fb6e4798e867080ba8115ad4cc6

**Review by alsofronie — APPROVED** [review]

File: src/Dropdown/Dropdown.jsx [hunk 7]
```diff
@@ -251,6 +322,31 @@ const Dropdown = ({
   )
 }
 
+const isEqual = (object1, object2) => {
```
Comment (sergiuelisei) [inline src/Dropdown/Dropdown.jsx:325]: I think this function should live in /utils

**Review by sergiuelisei — APPROVED** [review]

File: src/Dropdown/Dropdown.jsx (hunk 7 above)
Comment (hututudor) [reply src/Dropdown/Dropdown.jsx]: 605fde53ac953654bb1190e52d2d242b9fc3a9fd

**Pull request merged**